walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,162 @@
+    if (request.getAttendees().size() <= 0 && request.getOptionalAttendees().size() <= 0) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }  
+    
+    HashSet<String> mandatoryAttendees = new HashSet<String>(request.getAttendees());
+    HashSet<String> optionalAttendees = new HashSet<String>(request.getOptionalAttendees());
+    HashSet<String> allAttendees = Sets.newHashSet(Iterables.concat(mandatoryAttendees, optionalAttendees));
+      
+    String endOfDayTitle = "END_OF_DAY";
+    String startOfDayTitle = "START_OF_DAY";
+    Event endOfDay = new Event(endOfDayTitle, TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0), allAttendees);
+    Event startOfDay = new Event(startOfDayTitle, TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0), allAttendees);
+
+    ArrayList<Event> allEvents = new ArrayList<Event>(events);
+
+    allEvents.add(endOfDay);
+    allEvents.add(startOfDay);
+      
+    allEvents.sort(Event.EVENT_COMPARATOR);
+    
+    ArrayList<TimeRange> available = availableTimesFromSortedEvents(mandatoryAttendees, allEvents, request);
+
+    if (optionalAttendees.size() == 0) {
+      return available;
+    }
+
+    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedEvents(optionalAttendees, allEvents, request);
+    
+    if (available.size() == 0) {
+      return availableWithOptional;
+    }
+    
+    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+
+    for (TimeRange optionalTimeRange: availableWithOptional) {
+      int start = optionalTimeRange.start();
+      int end = optionalTimeRange.end();
+
+      int indexOfIntersection = binarySearchOverlappingTimeRange(available, 0, available.size() - 1, start);
+
+      if (indexOfIntersection != -1) {
+        int mandatoryEnd = available.get(indexOfIntersection).end();
+        if (end <= mandatoryEnd) {
+          finalTimeRanges.add(optionalTimeRange);
+        }
+      }
+    }
+
+    return finalTimeRanges.size() != 0 ? finalTimeRanges : available;
+  }
+    
+  public int binarySearchOverlappingTimeRange(ArrayList<TimeRange> mandatory, int startIndex, int endIndex, int startOfOptional) {
+    int startOfMandatory = mandatory.get(startIndex).start();
+
+    if (startIndex == endIndex) {
+      return startOfMandatory <= startOfOptional ? startIndex : -1;  
+    }
+
+    int midIndex = startIndex + (endIndex - startIndex) / 2;
+
+    if (startOfOptional < mandatory.get(midIndex).start()) {
+      return binarySearchOverlappingTimeRange(mandatory, startIndex, midIndex, startOfOptional);
+    }
+
+    int ret = binarySearchOverlappingTimeRange(mandatory, midIndex + 1, endIndex, startOfOptional);
+
+    return ret == -1 ? midIndex : ret;
+  }
+
+
+  /*
+   * Method to get available times from sorted events
+   * Does this by removing events where there's no attendees, then merging events
+   * which share some time. Finally the gaps are the result.
+   * 
+   * Runtime: O(n)
```

Comment:
Can you clarify which argument derives `n`? There are several collections passed into this method.